R/desc_utils.R

```diff
@@ -0,0 +1,201 @@
+#' Read DESCRIPTION file and return list of references.
+#'
+#' Returned list is an union between references specified in `"Config/Needs/verdepcheck"` field and
+#' standard references for all other not covered dependencies.
+#' @importFrom pkgdepends pkg_dep_types parse_pkg_ref
+#' @keywords internal
+#' @examplesIf Sys.getenv("R_USER_CACHE_DIR", "") != ""
+#' d <- desc::desc("!new")
+#' d$set_dep("foo", "Import")
+#' d$set_dep("bar", "Suggest")
+#' d$set_list("Config/Needs/verdepcheck", "foo/bar")
+#' d$set_list("Config/Needs/verdepcheck", "foo/baz") # not in package deps - will be skipped
+#' get_refs_from_desc(d)
+get_refs_from_desc <- function(d) {
+  if (.desc_field %nin% d$fields()) {
+    refs <- list()
+  } else {
+    refs <- lapply(get_desc_field_pkgs(d), pkgdepends::parse_pkg_ref)
+  }
+  all_deps <- d$get_deps()
+  all_deps <- all_deps$package[all_deps$type %in% pkgdepends::pkg_dep_types()]
+  missing_refs <- setdiff(
+    setdiff(all_deps, base_pkgs()),
+    map_key_character(refs, "package")
+  )
+  res <- c(
+    refs,
+    lapply(missing_refs, pkgdepends::parse_pkg_ref)
+  )
+  res_idx <- match(all_deps, map_key_character(res, "package"))
+  res_idx <- res_idx[!is.na(res_idx)]
+  res[res_idx]
+}
+
+#' Get the packages from the custom configuration field
+#' @param d (`desc`) DESCRIPTION object from [desc::description]
+#' @return character string
+#' @keywords internal
+get_desc_field_pkgs <- function(d) {
+  trimws(strsplit(d$get_field(.desc_field), ",")[[1]])
+}
+
+#' Replace Remotes in the `desc` that have been resolved to a GitHub tag or are
+#' in CRAN
+#'
+#' Replaces any existing Remotes entry with the resolved GitHub tag from
+#' `Config/Needs/verdepcheck`.
+#'
+#' @param d (`desc`) DESCRIPTION object
+#' @keywords internal
+#' @examples
+#' # Example that should replace rtables & formatters on Remotes
+#' #  but not pkgdepends
+#'
+#' d <- desc::desc(
+#'   file = verdepcheck:::local_description(
+#'     list(
+#'       rtables = "Import", formatters = "Import", pkgdepends = "Import",
+#'       dplyr = "Import"
+#'     ),
+#'     remotes = c(
+#'       "insightsengineering/rtables@*release",
+#'       "insightsengineering/formatters@*release",
+#'       "r-lib/pkgdepends@*release"
+#'     ),
+#'     need_verdepcheck = c(
+#'       "dplyr",
+#'       "rtables=insightsengineering/rtables@0.6.2",
+#'       "formatters=insightsengineering/formatters@0.5.1"
+#'     )
+#'   )
+#' )
+#' verdepcheck:::desc_remotes_cleanup(d)
+desc_remotes_cleanup <- function(d) {
+  # Parse the `Config/Needs/verdepcheck` to retrieve references and extract package names
+  desc_field_pkgs <- pkgdepends::parse_pkg_refs(get_desc_field_pkgs(d))
+  desc_field_pkg_names <- map_key_character(desc_field_pkgs, "package")
+
+  # Parse the remotes to retrieve the package names
+  remotes <- pkgdepends::parse_pkg_refs(d$get_remotes())
+  remotes_pkg <- map_key_character(remotes, "package")
```

Comment:
Can we please have some consistent var naming convention? It will be much easier to read.

For instance here we have `xxx_pkgs` and `yyy` indicating ref as a string and `xxx_pkg_names` and `yyy_pkg` as package name as a string. There is a conflict with `xxx_pkgs` and `yyy_pkg` storing different type of data.

Reply:
Sorry, missed that name, it should've been called `remotes_pkg_names`

Using `_refs` and `_names` 

https://github.com/insightsengineering/verdepcheck/blob/0819cbdbce04b16be2fe60f921d8677508d241d1/R/desc_utils.R#L80-L90